### PR TITLE
ref(rate limits): Use new style Django middleware

### PR DESCRIPTION
**Pull request by ceorourke**
Rewrite the rate limit middleware to not use the deprecated `MiddlewareMixin` - follows the pattern suggested in the Django docs [here](https://docs.djangoproject.com/en/4.0/topics/http/middleware/#writing-your-own-middleware).

### CODE REVIEW

**Review by AniketDas-Tekky — APPROVED** [review]

File: src/sentry/middleware/ratelimit.py
Comment (volokluev) [inline src/sentry/middleware/ratelimit.py:36]: since we're doing this change, I think it would be better to remove the `process_view` function and use the new middleware functionality in django more directly i.e.:

```python3
def __call__(self, request: Request):
    # whatever is done in `process_view` should be done here

    response = self.get_response(request)
    # whatever is done in `process_response` should be done here
```

That doesn't mean that having separate functions is bad but for me there are two reasons why what we have now is not ideal:

 1. Looking at this code, unless you know how django middleware works, it's not clear when `process_view` would ever be called. (I had to look it up myself because I had no idea how the tests weren't failing).
 2. There is some bad behavior we can clean up while we're here. You'll notice that we attach fields to the `request` object in `process_view` with the hope that they'll be there once we hit `process_response`. Example:
 
```python3
request.will_be_rate_limited = False
request.rate_limit_category = None
request.rate_limit_uid = uuid.uuid4().hex
request.rate_limit_key = get_rate_limit_key(view_func, request)
```
If these variables are in the same function scope, we can reduce the amount of baggage the request object needs to have. If we do that, the middleware component can be more self-contained and depend less on the global scope in the `request` object

Comment (AniketDas-Tekky) [reply src/sentry/middleware/ratelimit.py:36]: So we tried using the function approach yesterday. One problem is that we use data attached to the `view_func` for determining the rate limit, like the overrides. the `view_func` argument isn't available in the function middleware.

Comment (ceorourke) [reply src/sentry/middleware/ratelimit.py:36]: ^^ I tried again after reading this comment and got stuck w/o `view_func`, `get_rate_limit_key` relies on it. At the bare minimum what if I left a comment in `__call__` that says Django is calling `process_view`?

Comment (volokluev) [reply src/sentry/middleware/ratelimit.py:36]: It's up to you whether you want to do this. At the end of the day you're going to own this. In my opinion the code is is is a little clunky, specifically the rate limiter intreracttion

```
def process_view():
    metadata = above_rate_limit_check() # this checks the rate limiter and registers it in the concurrent rate limiter

def process_response():
    finish_request(request.rate_limit_key, request.uid) # this has to be called otherwise the concurrent slot will leak until cleanup is done
 ```
 
There is a more elegant way to do this if the request and response are in the same scope.
 
 Example pseudocode
 ```
 try:
     with rate_limit_scope(rate_limit_value):
        # do the request
 except RateLimitExceeded as e:
    return HTTPResponse(e.rate_limit_data)
```

However at the end of the day it's up to you what tradeoffs you want to make. If the way it is right now is ok with you, then it's ok with me

File: src/sentry/middleware/ratelimit.py
Comment (ceorourke) [inline src/sentry/middleware/ratelimit.py:39]: I pretty much only kept this in the `request` object so that we could still do `test_rate_limit_category`, otherwise I didn't know how to test it since we don't store that anywhere else.

File: tests/sentry/middleware/test_ratelimit_middleware.py
Comment (ceorourke) [inline tests/sentry/middleware/test_ratelimit_middleware.py:82]: Without `will_be_rate_limited` to assert on I went for the headers - it doesn't feel as accurate since sometimes `X-Sentry-Rate-Limit-Remaining` is 0 but that's fine since no more requests are coming, or sometimes it's 0 and that's not fine because another request is coming and it'd be rejected.

File: tests/sentry/middleware/test_ratelimit_middleware.py
Comment (ceorourke) [inline tests/sentry/middleware/test_ratelimit_middleware.py:304]: I had to take this out because we don't get to the code that adds the headers when there is an error response.

Comment (volokluev) [reply tests/sentry/middleware/test_ratelimit_middleware.py:304]: well that's a change in behavior right? Is there something we can do to not change this behavior?

File: tests/sentry/middleware/test_ratelimit_middleware.py
Comment (volokluev) [reply tests/sentry/middleware/test_ratelimit_middleware.py:82]: `will_be_rate_limited` was from a time when rate limits were not actually enforced. Now that they are, how can you determine that the request was rate limited?

File: tests/sentry/middleware/test_ratelimit_middleware.py
Comment (volokluev) [inline tests/sentry/middleware/test_ratelimit_middleware.py:70]: why is this test removed?

File: src/sentry/middleware/ratelimit.py
Comment (volokluev) [inline src/sentry/middleware/ratelimit.py:60]: is there a reason `rate_limit_metadata` still needs to be attached to the `request`?

File: src/sentry/middleware/ratelimit.py
Comment (volokluev) [inline src/sentry/middleware/ratelimit.py:41]: since this is slightly unusual, it may be worth leaving a comment as to why this step is being done

File: src/sentry/middleware/ratelimit.py
Comment (volokluev) [inline src/sentry/middleware/ratelimit.py:36]: what's the point of this `data` variable?

**Review by volokluev — COMMENTED** [review]
some comments and questions

File: tests/sentry/middleware/test_ratelimit_middleware.py
Comment (ceorourke) [reply tests/sentry/middleware/test_ratelimit_middleware.py:82]: Do we actually need to know that outside of tests? It's not used anywhere else.

File: src/sentry/middleware/ratelimit.py
Comment (ceorourke) [reply src/sentry/middleware/ratelimit.py:36]: I guess it doesn't need to be in a dictionary, but I was following up on a TODO left from before that said to put the fields in an object. I have to initialize it here to get around an unbound variable error.

Comment (volokluev) [reply src/sentry/middleware/ratelimit.py:36]: ah yes, that was at the time when we needed to attach it to the request object. This way all the things that were needed to be bubbled up to other places outside the middleware were all under one namespace and not scattered around. If you're not attaching these things to the request, then there's no reason to put them in an object.

File: tests/sentry/middleware/test_ratelimit_middleware.py
Comment (volokluev) [reply tests/sentry/middleware/test_ratelimit_middleware.py:82]: > Do we actually need to know that outside of tests? 
It depends on what you mean by "that". As you pointed out, using `X-Sentry-Rate-Limit-Remaining` is not sufficient to tell us if it the request was rate limited (and `will_be_rate_limited` was). But we're testing the rate limit middleware, so presumably we should test to see that it does in fact rate limit.

> It's not used anywhere else.
Are you sure?

Comment (ceorourke) [reply tests/sentry/middleware/test_ratelimit_middleware.py:82]: Arg it's used in the access log. What I'm stuck on is that we are trying to not add things to the request object but we also want extra things on the request object like `will_be_rate_limited` and `rate_limit_category`. I tried returning 2 things from `__call__` with bad results earlier.

File: tests/sentry/middleware/test_ratelimit_middleware.py
Comment (ceorourke) [reply tests/sentry/middleware/test_ratelimit_middleware.py:70]: I removed it since we have `test_fails_open` and we no longer directly call `process_response` ... but the fact that you're commenting on it makes me feel like I made a mistake.

File: src/sentry/middleware/ratelimit.py
Comment (ceorourke) [reply src/sentry/middleware/ratelimit.py:60]: Pretty much just because I can't figure out how else to get it into the access log.

File: src/sentry/middleware/ratelimit.py
Comment (volokluev) [inline src/sentry/middleware/ratelimit.py:44]: nit: this should be one line above the "unusual" call to resolve. Otherwise it seems like the comment is about the `get_rate_limit_key` function

File: src/sentry/middleware/ratelimit.py
Comment (volokluev) [inline src/sentry/middleware/ratelimit.py:47]: you call `get_response` three times in this function. This makes there be more moving parts and places to mess up for future programmers. Is there a way to re-organize the code so that you only have to call it once?

File: src/sentry/middleware/ratelimit.py
Comment (volokluev) [inline src/sentry/middleware/ratelimit.py:79]: You have to remember to call `self.add_headers` twice to make this code work properly. Can it be reorganized to only be called once so that people modifying this code in the future don't have to remember to?

File: src/sentry/middleware/ratelimit.py
Comment (volokluev) [inline src/sentry/middleware/ratelimit.py:95]: In this case, if the `rate_limit_metadata` is not present this will log an exception. But the rate_limit_metadata may not be present simply because the endpoint is not supposed to be rate limited. Exceptions should only be logged when they are actually exceptions.

**Review by volokluev — APPROVED** [review]
I'm approving with the trust that you'll resolve the following things:

1. The `logging.exception` should not happen anymore because there's no exception
2. You will at least consider what I said about the repeated calls to `get_response` and `add_headers`. Whatever you decide, understanding why you made the decision is the important part.

**Comment by volokluev** [comment]
Also, probably don't ship this until Monday. A mistake here can take down the whole API, so you want to be around for that and monitor your deploy in sentry

File: src/sentry/middleware/ratelimit.py
Comment (volokluev) [inline src/sentry/middleware/ratelimit.py:48]: Since there's an effort to type the sentry codebase right now, let's be good citizens and make this function return a dataclass as opposed to a dict[Any, Any]. That way we can be more explicit about what this function is returning and the typechecker can do more work for us.

**Review by volokluev — APPROVED** [review]
Looks good, just one thing on the typing

**Review by volokluev — APPROVED** [review]

**Pull request merged**
